platform: nordic_nrf: Add CMSIS USART driver

anangl opened the pull request:
Add an implementation of the CMSIS USART driver API that uses
the nrfx_uarte driver underneath. Plus corresponding configuration
files for nRF5340 and nRF9160, and related translation layer that
activates relevant parts of nrfx.

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,341 @@
+    ARM_USART_DRV_VERSION
+};
+
+static const ARM_USART_CAPABILITIES DriverCapabilities = {
```
Just curious: Which of these will we eventually support?

anangl replied:
I used here what is currently used by all other platforms (and what is actually just copied from the driver template). TF-M code does not seem to be checking any of these capabilities. I thought about signalling RTS and CTS flow control as available, but the problem is that these two are separate capabilities here and our hardware could not support an RTS-only flow control, which is defined by the Driver_USART API.
I'll leave it as it is now.

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,341 @@
+{
+    nrfx_err_t err_code = nrfx_uarte_tx(&uart_resources->uarte, data, num);
+    if (err_code != NRFX_SUCCESS) {
+        return ARM_DRIVER_ERROR_BUSY;
```
This might hide more serious errors, like NRFX_ERROR_INTERNAL.

anangl replied:
Right. Will change this.

oyvindronningstad replied:
You could create a macro and use it everywhere? I see initialize is the same way.

oyvindronningstad replied:
Should ARM_USARTx_Initialize() get the same treatment?

anangl replied:
No, the only error in `ARM_USARTx_Initialize` can be that there is an attempt to initialize an already initialized nrfx_uarte driver instance.
In `ARM_USARTx_Send` and `ARM_USARTx_Receive`, the driver can be busy with a previous transfer (`ARM_DRIVER_ERROR_BUSY` is returned then) or some other error can occur, like invalid buffer passed (not placed in RAM) or an internal nrfx_uarte driver error, then `ARM_DRIVER_ERROR` is returned.

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,341 @@
+        return ARM_DRIVER_ERROR_BUSY;
+    }
+
+    uart_resources->tx_count = num;
```
Will `uart_resources->tx_count` always be 0 before this? Otherwise, should this be `+=`?

anangl replied:
This is what `ARM_USARTx_GetTxCount` returns and the description for that function [says](https://github.com/NordicPlayground/trusted-firmware-m/blob/master/platform/ext/driver/Driver_USART.h#L252) it should return the "number of data items transmitted". Other platforms return the number of bytes sent in the last transaction (i.e. call to `ARM_USART_Send`). I used the same approach.

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,341 @@
+        return ARM_DRIVER_ERROR_UNSUPPORTED;
+    }
+
+    nrfx_uarte_config_t *config = &uart_resources->config;
```
If the function exits early, this config will not match the actual configuration since the uninit/init didn't run. WIll that be a problem?

anangl replied:
I thought about it. It should not be, as it is actually not described how the function should behave when only some of requested changes are successful. But using a local copy of the `config` structure seems to be safer. I'll rework it.

anangl committed:
platform: nordic_nrf: Fix definition of NRFX_DELAY_US

Use `nrfx_coredep_delay_us` instead of non-existing `nrfx_coredep_us`.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Dropped [DNM] commits and rebased on master.

ioannisg commented:
> Dropped [DNM] commits and rebased on master.

Strill in draft mode, AFAICS, @anangl

anangl committed:
platform: nordic_nrf: Add CMSIS USART driver

Add an implementation of the CMSIS USART driver API that uses
the nrfx_uarte driver underneath. Plus corresponding configuration
files for nRF5340 and nRF9160, and related translation layer that
activates relevant parts of nrfx.

Signed-off-by: Andrzej Głąbek <andrzej.glabek@nordicsemi.no>

anangl commented:
Refactored the code a bit, mostly `ARM_USARTx_Control()`, trying to address @oyvindronningstad's comments.
Added also a fix for an issue in `nrfx_glue.h` revealed while building this driver with the updated nrfx.

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,363 @@
+
+        case ARM_USART_PARITY_EVEN:
+            hal_cfg.parity = NRF_UARTE_PARITY_INCLUDED;
+            hal_cfg.paritytype = NRF_UARTE_PARITYTYPE_EVEN;
```
I get a compile error here for nrf9160. Seems it doesn't have UARTE_CONFIG_PARITYTYPE_Msk.

anangl replied:
Right, I tested only on nRF5340. My bad. Will fix it promptly.

anangl replied:
Fixed.

oyvindronningstad commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,363 @@
+
+        case ARM_USART_PARITY_ODD:
+            hal_cfg.parity = NRF_UARTE_PARITY_INCLUDED;
+            hal_cfg.paritytype = NRF_UARTE_PARITYTYPE_ODD;
```
And here

ioannisg commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,369 @@
+    case ARM_POWER_FULL:
+        /* Nothing to be done */
+        return ARM_DRIVER_OK;
+    /* default:  The default is not defined intentionally to force the
```
Several drivers define the default case to force the function return DRIVER_ERROR_PARAMETER - I wonder if you thought of it an chose to comply with targets that do nothing about it.

anangl replied:
I just took the approach that has been used by majority of platforms in this regard. But now I see that in Flash drivers `ARM_DRIVER_UNSUPPORTED` is mostly used in this case. This code actually seems to be a better fit than `ARM_DRIVER_ERROR_PARAMETER`. And this is also what is used in driver templates in CMSIS. I will switch to that one.

ioannisg commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,369 @@
+    return DriverCapabilities;
+}
+
+int32_t ARM_USARTx_Initialize(ARM_USART_SignalEvent_t cb_event,
```
@anangl why these functions are not defined as static?

anangl replied:
Too much relaying on the driver template. Obviously, they should be `static`. I realized that while preparing the Flash driver.

ioannisg reviewed (commented):
Looks fine to me.
I have a question about the driver functions not been defined with local scope, I wonder if I am missing something here.

ioannisg commented on platform/ext/target/nordic_nrf/common/cmsis_drivers/Driver_USART.c:
```diff
@@ -0,0 +1,369 @@
+    }
+
+    uart_resources->tx_count = num;
+    return NRFX_SUCCESS;
```
I wonder whether this is intentional or overlooked: some function return codes are taken from nrfX, such as NRFX_SUCCESS, while in other cases TF-M error code definitions are used.

anangl replied:
Overlooked. Should be `ARM_DRIVER_OK`. Will fix it.

anangl commented:
- marked all functions that don't need to be available externally as `static`
- simplified initialization of `DriverCapabilities`
- refactored `ARM_USARTx_PowerControl`
- corrected values returned by `ARM_USARTx_Send` and `ARM_USARTx_Receive`

ioannisg reviewed (approved):
thanks for the updates @anangl

ioannisg commented:
@oyvindronningstad would you want/need to take another look here?

oyvindronningstad reviewed (approved):
LGTM
